### DISCUSSION TITLE
Teensy 4.0 > Sprite > Colors wrong

### DISCUSSION THREAD

**Discussion by ZL1CVD** [Bodmer/TFT_eSPI#2753] [Q&A - Sprites]
I'm having color problems with Teensy 4.0 & sprites only - everything works great with direct write.
I int sprite with:
`sprR.setColorDepth(16); `    
`sprR.setSwapBytes(true);`
`sprR.createSprite(196, 75); ` 

then push sprite with:

`if  (fft1024R.available() == true) {`
`sprR.fillSprite(TFT_BLACK);`
`for (int i = 1; i < 182; i++) { `
`int index = i / 2 + 1;`
`int dta = constrain((fft1024R.read(index) + fft1024R.read(index + 1)) * 256.0, 0, 31); `
`sprR.drawLine(i + 6, 32 + dta, i + 6, 32 - dta, TFT_BLUE); `
`}`
`sprR.pushSprite(_c2, _r1);`
` }`
TFT_BLUE comes out as TFT_GREEN. In other tests TFT_GREEN comes out as  TFT_RED

Changing setSwapBytes() has no effect. I suspect its something to do with the Teensy Arm chip being handled as generic in the library.

**Comment by noorino1977** [comment]
#include <TFT_eSPI.h>                 // Include the graphics library (this includes the sprite functions)

TFT_eSPI    tft = TFT_eSPI();         // Create object "tft"
TFT_eSprite waterfall = TFT_eSprite(&tft);  // Create Sprite object "img" with pointer to "tft" object
 
int colourcheck1 = 0;
int colourcheck2 = 1000;
int colourcheck3 = 2000;
int colourcheck4 = 3000;
int colourcheck5 = 4000;
int colourcheck6 = 5000;
int colourcheck7 = 6000;
int colourcheck8 = 7000;
int colourcheck9 = 0;
   

int coY=0;
void setup(void) {
  tft.init();
  tft.setRotation(1);
  tft.fillScreen(TFT_BLUE);
  waterfall.setColorDepth(8);  waterfall.createSprite(480, 320);  waterfall.fillSprite(TFT_BLACK);
  
}

// -------------------------------------------------------------------------
// Main loop
// -------------------------------------------------------------------------
void loop() {

    for (int x = 0 ; x<10000; x+=2 ){
         //waterfall.fillCircle(x, coY, 3,colourcheck); 
         waterfall.pushSprite(0, 0);
         
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck1), 110 , 10, 2); waterfall.fillCircle(170, 10, 8,colourcheck1); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck2), 110 , 40, 2); waterfall.fillCircle(170, 40, 8, colourcheck2);     
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck3), 110 , 80, 2); waterfall.fillCircle(170, 80, 8, colourcheck3); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck4), 110 , 120, 2); waterfall.fillCircle(170, 120, 8, colourcheck4); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck5), 110 , 160, 2); waterfall.fillCircle(170, 160, 8, colourcheck5); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck6), 110 , 200, 2); waterfall.fillCircle(170, 200, 8, colourcheck6);  
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck7), 110 , 240, 2); waterfall.fillCircle(170, 240, 8, colourcheck6);  
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck8), 110 , 280, 2); waterfall.fillCircle(170, 280, 8, colourcheck6);       
 colourcheck1++;
 colourcheck2++;
 colourcheck3++;
 colourcheck4++;
 colourcheck5++;
 colourcheck6++;
 colourcheck7++;
 colourcheck8++;
         delay(100);
    }
         //sTFT_BLACK=,     
         //sTFT_BLUE=, 
         //sTFT_YELLOW=,
         //sTFT_RED=, 

}

**Reply by noorino1977** [reply]
the same as my problem, i have do a crazy one to find the colour is

**Reply by noorino1977** [reply]
#include <TFT_eSPI.h>                 // Include the graphics library (this includes the sprite functions)

TFT_eSPI    tft = TFT_eSPI();         // Create object "tft"
TFT_eSprite waterfall = TFT_eSprite(&tft);  // Create Sprite object "img" with pointer to "tft" object
 
int kali=10;
int colourcheck1 = 0*kali;
int colourcheck2 = 1000*kali;
int colourcheck3 = 2000*kali;
int colourcheck4 = 3000*kali;
int colourcheck5 = 4000*kali;
int colourcheck6 = 5000*kali;
int colourcheck7 = 6000*kali;
int colourcheck8 = 7000*kali;
int colourcheck9 = 8000*kali;
int colourcheck10 = 9000*kali;   

int coY=0;
void setup(void) {
  tft.init();
  tft.setRotation(2);
  tft.fillScreen(TFT_BLUE);
  waterfall.setColorDepth(8);  waterfall.createSprite(320, 480);  waterfall.fillSprite(TFT_BLACK);
  
}

// -------------------------------------------------------------------------
// Main loop
// -------------------------------------------------------------------------
void loop() {

    for (int x = 0 ; x<10000; x+=2 ){
         //waterfall.fillCircle(x, coY, 3,colourcheck); 
         waterfall.pushSprite(0, 0);
         
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck1), 40 , 10, 2); waterfall.fillCircle(80, 10, 8,colourcheck1); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck2), 40 , 40, 2); waterfall.fillCircle(80, 40, 8, colourcheck2);     
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck3), 40 , 80, 2); waterfall.fillCircle(80, 80, 8, colourcheck3); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck4), 40 , 120, 2); waterfall.fillCircle(80, 120, 8, colourcheck4); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck5), 40 , 160, 2); waterfall.fillCircle(80, 160, 8, colourcheck5); 
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck6), 40 , 200, 2); waterfall.fillCircle(80, 200, 8, colourcheck6);  
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck7), 40 , 240, 2); waterfall.fillCircle(80, 240, 8, colourcheck7);  
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck8), 40 , 280, 2); waterfall.fillCircle(80, 280, 8, colourcheck8);      
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck9), 40 , 320, 2); waterfall.fillCircle(80, 320, 8, colourcheck9);  
         waterfall.setTextColor(TFT_WHITE,TFT_BLACK);  waterfall.drawCentreString(String(colourcheck10), 40 , 360, 2); waterfall.fillCircle(80, 360, 8, colourcheck10);  
 
 
 colourcheck1++;
 colourcheck2++;
 colourcheck3++;
 colourcheck4++;
 colourcheck5++;
 colourcheck6++;
 colourcheck7++;
 colourcheck8++;
 colourcheck9++;
 colourcheck10++;



         delay(500);
    }
         //sTFT_BLACK=,     
         //sTFT_BLUE=60010, 
         //sTFT_YELLOW=1950,
         //sTFT_RED=1790, 

}
